File: src/calculus_optimization_hard.py
"""
calculus - optimization (hard)
Generated: 2026-02-22T05:09:48.741977
"""

from problem_utils import *

def generate():
    problem_type = choice(['box', 'fence', 'cylinder', 'distance', 'area_triangle', 'window'])
    
    if problem_type == 'box':
        # Open box from cutting corners of rectangular sheet
        length = randint(8, 20)
        width = randint(8, 20)
        
        # Volume function V(x) = x(L-2x)(W-2x)
        V = x * (length - 2*x) * (width - 2*x)
        V_expanded = expand(V)
        
        # Find critical points
        dV = diff(V, x)
        critical_pts = solve(dV, x)
        
        # Filter for valid dimensions (0 < x < min(L,W)/2)
        valid_crit = [pt for pt in critical_pts if pt.is_real and pt > 0 and pt < min(length, width)/2]
        
        if valid_crit:
            x_max = valid_crit[0]
            V_max = V.subs(x, x_max)
            
            return problem(
                question=f"A rectangular sheet of metal measuring ${length}$ cm by ${width}$ cm has squares of side length $x$ cut from each corner. The sides are then folded up to form an open box. Find the value of $x$ that maximizes the volume of the box.",
                answer=x_max,
                difficulty=(1650, 1750),
                topic="calculus/optimization",
                solution=steps(
                    f"The box has dimensions: length = ${length - 2}x$, width = ${width - 2}x$, height = $x$",
                    f"Volume: $V(x) = x({length} - 2x)({width} - 2x) = {latex(V_expanded)}$",
                    f"Take derivative: $V'(x) = {latex(dV)}$",
                    f"Set $V'(x) = 0$ and solve: ${latex(Eq(dV, 0))}$",
                    f"Critical point: $x = {latex(x_max)}$",
                    f"Verify this is a maximum by checking $V''(x)$ or endpoints.",
                    f"The value that maximizes volume is $x = {latex(x_max)}$ cm"
                ),
                calculator="scientific"
            )
    
    elif problem_type == 'fence':
        # Rectangular enclosure with one side against building
        area = choice([800, 1200, 1600, 2000, 2400])
        
        # Area = xy = A, so y = A/x
        # Perimeter to minimize: P = x + 2y = x + 2A/x
        P = x + 2*area/x
        dP = diff(P, x)
        
        x_min = solve(dP, x)[0]
        y_min = area / x_min
        P_min = P.subs(x, x_min)
        
        return problem(
            question=f"A rectangular enclosure is to be built against an existing wall, requiring fencing on only three sides. If the enclosure must have an area of ${area}$ square meters, find the dimensions that minimize the amount of fencing needed. What is the length of the side parallel to the wall?",
            answer=x_min,
            difficulty=(1600, 1700),
            topic="calculus/optimization",
            solution=steps(
                f"Let $x$ = length parallel to wall, $y$ = width perpendicular to wall",
                f"Constraint: $xy = {area}$, so $y = \\frac{{{area}}}{{x}}$",
                f"Minimize perimeter: $P(x) = x + 2y = x + \\frac{{{2*area}}}{{x}}$",
                f"Take derivative: $P'(x) = {latex(dP)}$",
                f"Set $P'(x) = 0$: $1 - \\frac{{{2*area}}}{{x^2}} = 0$",
                f"Solve: $x^2 = {2*area}$, so $x = {latex(x_min)}$ meters",
                f"Then $y = \\frac{{{area}}}{{{latex(x_min)}}} = {latex(y_min)}$ meters",
                f"The side parallel to the wall is ${latex(x_min)}$ meters"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'cylinder':
        # Cylinder with fixed surface area, maximize volume
        surface_area = choice([100, 200, 300, 400]) * pi
        SA_coef = surface_area / pi
        
        # Surface area: 2πr² + 2πrh = SA
        # So h = (SA - 2πr²)/(2πr)
        # Volume V = πr²h = πr² · (SA - 2πr²)/(2πr) = r(SA - 2πr²)/2
        
        h_expr = (surface_area - 2*pi*x**2) / (2*pi*x)
        V = pi * x**2 * h_expr
        V_simplified = simplify(V)
        
        dV = diff(V_simplified, x)
        r_max = solve(dV, x)[0]
        h_max = h_expr.subs(x, r_max)
        
        return problem(
            question=f"A closed cylindrical can has a surface area of ${latex(surface_area)}$ square units. Find the radius that maximizes the volume of the can.",
            answer=r_max,
            difficulty=(1700, 1800),
            topic="calculus/optimization",
            solution=steps(
                f"Let $r$ = radius, $h$ = height",
                f"Surface area: $2\\pi r^2 + 2\\pi rh = {latex(surface_area)}$",
                f"Solve for $h$: $h = \\frac{{{latex(surface_area)} - 2\\pi r^2}}{{2\\pi r}}$",
                f"Volume: $V = \\pi r^2 h = {latex(V_simplified)}$",
                f"Take derivative: $V'(r) = {latex(dV)}$",
                f"Set $V'(r) = 0$ and solve",
                f"Optimal radius: $r = {latex(r_max)}$ units"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'distance':
        # Point on curve closest to given point
        px = randint(3, 8)
        py = 0
        
        # Use parabola y = x²
        # Distance squared: D² = (x - px)² + (x² - py)²
        D_squared = (x - px)**2 + (x**2 - py)**2
        
        dD = diff(D_squared, x)
        critical = solve(dD, x)
        
        # Find real critical point closest to px
        real_crit = [pt for pt in critical if pt.is_real]
        if real_crit:
            x_closest = min(real_crit, key=lambda pt: abs(float(pt - px)))
            y_closest = x_closest**2
            min_dist = sqrt(D_squared.subs(x, x_closest))
            
            return problem(
                question=f"Find the point on the parabola $y = x^2$ that is closest to the point $({px}, {py})$. What is the $x$-coordinate of this point?",
                answer=x_closest,
                difficulty=(1750, 1850),
                topic="calculus/optimization",
                solution=steps(
                    f"Point on parabola: $(x, x^2)$. Given point: $({px}, {py})$",
                    f"Distance squared: $D^2 = (x - {px})^2 + (x^2 - {py})^2 = {latex(expand(D_squared))}$",
                    f"Minimize by taking derivative: $\\frac{{dD^2}}{{dx}} = {latex(dD)}$",
                    f"Set equal to zero and solve: ${latex(Eq(dD, 0))}$",
                    f"Critical point: $x = {latex(x_closest)}$",
                    f"Verify this is a minimum using the second derivative test"
                ),
                calculator="scientific"
            )
    
    elif problem_type == 'area_triangle':
        # Isosceles triangle with fixed perimeter, maximize area
        perimeter = randint(15, 30)
        
        # Let base = x, two equal sides = s each
        # Perimeter: x + 2s = P, so s = (P - x)/2
        # Height by Pythagorean: h² = s² - (x/2)²
        # Area = (1/2)xh
        
        s_expr = (perimeter - x) / 2
        h_squared = s_expr**2 - (x/2)**2
        h_squared_simplified = simplify(h_squared)
        
        # Area² = (x²/4)(s² - x²/4)
        A_squared = (x**2 / 4) * h_squared_simplified
        A_squared_expanded = expand(A_squared)
        
        dA = diff(A_squared_expanded, x)
        x_max = solve(dA, x)[0]
        
        return problem(
            question=f"An isosceles triangle has a perimeter of ${perimeter}$ units. Find the length of the base that maximizes the area of the triangle.",
            answer=x_max,
            difficulty=(1700, 1800),
            topic="calculus/optimization",
            solution=steps(
                f"Let base = $x$, equal sides = $s$ each",
                f"Perimeter constraint: $x + 2s = {perimeter}$, so $s = \\frac{{{perimeter} - x}}{{2}}$",
                f"Height: $h = \\sqrt{{s^2 - \\frac{{x^2}}{{4}}}}$",
                f"Area: $A = \\frac{{1}}{{2}}xh = \\frac{{x}}{{2}}\\sqrt{{s^2 - \\frac{{x^2}}{{4}}}}$",
                f"To maximize, consider $A^2 = \\frac{{x^2}}{{4}}\\left(s^2 - \\frac{{x^2}}{{4}}\\right) = {latex(A_squared_expanded)}$",
                f"Take derivative: $\\frac{{d(A^2)}}{{dx}} = {latex(dA)}$",
                f"Set equal to zero and solve: $x = {latex(x_max)}$ units"
            ),
            calculator="scientific"
        )
    
    else:  # window
        # Norman window (rectangle with semicircle on top), fixed perimeter
        perimeter = randint(12, 24)
        
        # Let width = 2r (diameter of semicircle), height of rectangle = h
        # Perimeter: 2h + 2r + πr = P
        # So h = (P - 2r - πr)/2
        # Area: 2rh + (1/2)πr²
        
        h_expr = (perimeter - 2*x - pi*x) / 2
        A = 2*x*h_expr + pi*x**2/2
        A_simplified = simplify(A)
        
        dA = diff(A_simplified, x)
        r_max = solve(dA, x)[0]
        
        return problem(
            question=f"A Norman window consists of a rectangle surmounted by a semicircle. If the perimeter of the window is ${perimeter}$ meters, find the radius of the semicircle that maximizes the area of the window.",
            answer=r_max,
            difficulty=(1800, 1900),
            topic="calculus/optimization",
            solution=steps(
                f"Let $r$ = radius of semicircle (so width = $2r$), $h$ = height of rectangle",
                f"Perimeter: $2h + 2r + \\pi r = {perimeter}$",
                f"Solve for $h$: $h = \\frac{{{perimeter} - 2r - \\pi r}}{{2}}$",
                f"Area: $A = 2rh + \\frac{{1}}{{2}}\\pi r^2 = {latex(A_simplified)}$",
                f"Take derivative: $A'(r) = {latex(dA)}$",
                f"Set $A'(r) = 0$ and solve",
                f"Optimal radius: $r = {latex(r_max)}$ meters"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/calculus_optimization_hard.jl
# calculus - optimization (hard)
# Generated: 2026-03-08T20:43:33.489151

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("calculus/optimization")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Box optimization with fixed perimeter
        perimeter = randint(40, 120)
        optimal_x = perimeter // 4
        optimal_y = perimeter // 4
        max_area = optimal_x * optimal_y
        
        problem(
            question="A rectangle has a fixed perimeter of $(perimeter) units. Find the dimensions that maximize the area.",
            answer=fmt_tuple((optimal_x, optimal_y)),
            difficulty=(1800, 2000),
            answer_type="tuple",
            solution=steps(
                "Let x and y be the dimensions of the rectangle",
                sol("Constraint", 2x + 2y ~ perimeter),
                sol("Area", "A = xy"),
                "Express y in terms of x: \$y = $(perimeter//2) - x\$",
                sol("A(x)", "A = x($(perimeter//2) - x) = $(perimeter//2)x - x^2"),
                sol("\\frac{dA}{dx}", "$(perimeter//2) - 2x"),
                "Set derivative to zero: \$$(perimeter//2) - 2x = 0\$",
                sol("Critical point", "x = $(optimal_x)"),
                "Therefore \$y = $(optimal_y)\$",
                "Second derivative test: \$\\frac{d^2A}{dx^2} = -2 < 0\$ confirms maximum",
                sol("Dimensions", fmt_tuple((optimal_x, optimal_y)))
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Cylinder optimization - maximize volume with fixed surface area
        S = randint(200, 800)
        r_opt = sqrt(S // (6 * π))
        h_opt = 2 * r_opt
        V_max = π * r_opt^2 * h_opt
        
        problem(
            question="A cylindrical can has a fixed surface area of $(S)π square units. Find the radius that maximizes the volume. (Surface area formula: \$S = 2\\pi r^2 + 2\\pi rh\$)",
            answer=r_opt,
            difficulty=(2000, 2200),
            solution=steps(
                sol("Given", "S = 2\\pi r^2 + 2\\pi rh = $(S)\\pi"),
                "Solve for h: \$h = \\frac{$(S)\\pi - 2\\pi r^2}{2\\pi r} = \\frac{$(S) - 2r^2}{2r}\$",
                sol("Volume", "V = \\pi r^2 h = \\pi r^2 \\cdot \\frac{$(S) - 2r^2}{2r} = \\frac{\\pi r($(S) - 2r^2)}{2}"),
                "Expand: \$V = \\frac{$(S)\\pi r - 2\\pi r^3}{2}\$",
                sol("\\frac{dV}{dr}", "\\frac{$(S)\\pi - 6\\pi r^2}{2}"),
                "Set derivative to zero: \$$(S)\\pi - 6\\pi r^2 = 0\$",
                sol("Critical point", "r^2 = $(S//6), \\quad r = \\sqrt{\\frac{$(S)}{6}}"),
                "Second derivative: \$\\frac{d^2V}{dr^2} = -6\\pi r < 0\$ confirms maximum",
                sol("Optimal radius", "r = \\sqrt{\\frac{$(S)}{6\\pi}}")
            ),
            time=240
        )
        
    elseif problem_type == 3
        # Fence optimization with one side against wall
        area_target = randint(400, 1600)
        optimal_x = sqrt(2 * area_target)
        optimal_y = area_target // optimal_x
        min_fence = optimal_x + 2 * optimal_y
        
        problem(
            question="A farmer wants to enclose a rectangular area of $(area_target) square meters using fencing on three sides (one side is against an existing wall). Find the dimensions that minimize the amount of fencing needed.",
            answer=fmt_tuple((optimal_x, optimal_y)),
            difficulty=(1900, 2100),
            answer_type="tuple",
            solution=steps(
                "Let x = width parallel to wall, y = depth perpendicular to wall",
                sol("Constraint", "xy = $(area_target)"),
                sol("Fence length", "F = x + 2y"),
                "Express y in terms of x: \$y = \\frac{$(area_target)}{x}\$",
                sol("F(x)", "F = x + 2\\cdot\\frac{$(area_target)}{x} = x + \\frac{$(2*area_target)}{x}"),
                sol("\\frac{dF}{dx}", "1 - \\frac{$(2*area_target)}{x^2}"),
                "Set derivative to zero: \$1 - \\frac{$(2*area_target)}{x^2} = 0\$",
                sol("Solve", "x^2 = $(2*area_target), \\quad x = \\sqrt{$(2*area_target)}"),
                "Therefore \$y = \\frac{$(area_target)}{\\sqrt{$(2*area_target)}} = \\sqrt{\\frac{$(area_target)}{2}}\$",
                "Second derivative: \$\\frac{d^2F}{dx^2} = \\frac{$(4*area_target)}{x^3} > 0\$ confirms minimum",
                sol("Dimensions", "(x, y) = (\\sqrt{$(2*area_target)}, \\sqrt{$(area_target//2)})")
            ),
            time=200
        )
        
    elseif problem_type == 4
        # Distance optimization - point to parabola
        a = nonzero(-3, 3)
        b = randint(-8, 8)
        c = randint(-12, 12)
        px = randint(-10, 10)
        py = randint(-10, 10)
        
        # Critical point from calculus
        crit_x = (2*a*px - b) // (2*a)
        crit_y = a*crit_x^2 + b*crit_x + c
        min_dist = sqrt((crit_x - px)^2 + (crit_y - py)^2)
        
        problem(
            question="Find the point on the parabola \$y = $(tex(a*x^2 + b*x + c))\$ that is closest to the point \$($(px), $(py))\$.",
            answer=fmt_tuple((crit_x, crit_y)),
            difficulty=(2100, 2300),
            answer_type="tuple",
            solution=steps(
                "Let \$(x, y)\$ be a point on the parabola where \$y = $(tex(a*x^2 + b*x + c))\$",
                sol("Distance squared", "D^2 = (x - $(px))^2 + (y - $(py))^2"),
                "Substitute parabola equation:",
                sol("D^2(x)", "(x - $(px))^2 + ($(tex(a*x^2 + b*x + c)) - $(py))^2"),
                "Take derivative and set to zero (using chain rule)",
                "After differentiation: \$\\frac{d(D^2)}{dx} = 2(x - $(px)) + 2($(tex(a*x^2 + b*x + c)) - $(py))\\cdot($(tex(diff(a*x^2 + b*x + c, x))))\$",
                "Set equal to zero and solve for critical points",
                sol("Critical x-value", "x = $(crit_x)"),
                sol("Corresponding y-value", "y = $(crit_y)"),
                sol("Closest point", fmt_tuple((crit_x, crit_y)))
            ),
            time=280
        )
        
    else
        # Norman window optimization (semicircle on top of rectangle)
        perimeter = randint(16, 36)
        # Optimal: r = P/(4+π)
        r_opt = perimeter // (4 + π)
        h_opt = (perimeter - π*r_opt) // 2 - r_opt
        area_max = 2*r_opt*h_opt + π*r_opt^2 // 2
        
        problem(
            question="A Norman window has the shape of a rectangle surmounted by a semicircle. If the perimeter of the window is $(perimeter) meters, find the radius of the semicircle that maximizes the area.",
            answer=r_opt,
            difficulty=(2200, 2400),
            solution=steps(
                "Let r = radius of semicircle, h = height of rectangle",
                sol("Perimeter", "P = 2h + 2r + \\pi r = $(perimeter)"),
                "Solve for h: \$h = \\frac{$(perimeter) - 2r - \\pi r}{2} = \\frac{$(perimeter)}{2} - r - \\frac{\\pi r}{2}\$",
                sol("Area", "A = 2rh + \\frac{\\pi r^2}{2}"),
                "Substitute h:",
                sol("A(r)", "A = 2r\\left(\\frac{$(perimeter)}{2} - r - \\frac{\\pi r}{2}\\right) + \\frac{\\pi r^2}{2}"),
                "Expand: \$A = $(perimeter)r - 2r^2 - \\pi r^2 + \\frac{\\pi r^2}{2} = $(perimeter)r - 2r^2 - \\frac{\\pi r^2}{2}\$",
                sol("\\frac{dA}{dr}", "$(perimeter) - 4r - \\pi r"),
                "Set derivative to zero: \$$(perimeter) - 4r - \\pi r = 0\$",
                sol("Solve", "r(4 + \\pi) = $(perimeter)"),
                sol("Optimal radius", "r = \\frac{$(perimeter)}{4 + \\pi}"),
                "Second derivative: \$\\frac{d^2A}{dr^2} = -4 - \\pi < 0\$ confirms maximum"
            ),
            time=300
        )
    end
end